### DISCUSSION TITLE
Porting layered material implementations from Mitsuba 2

### DISCUSSION THREAD

**Discussion by joeylitalien** [mitsuba-renderer/mitsuba3#845] [General]
Hi all,

I have been looking for a true implementation of layered materials in Mitsuba 3 but the closest I could find seems to be through the principled BRDF which isn't really layered. Both [Tizian's paper](https://rgl.epfl.ch/publications/Zeltner2018Layer) and the more recent work on [pearlescent materials](https://rgl.epfl.ch/publications/Guillen2020Pearlescent) are implemented in Mitsuba 2. Generally speaking, how hard would it be to convert this code to Mitsuba 3? This is not really a feature request; I'm mostly asking how feasible it would be to do this conversion myself. Any pitfalls I should be aware of if I were to go ahead?

Keep up the amazing work, you are doing God's work with the development of Mitsuba 3.

Cheers,
Joey

**Comment by tizian** [comment]
Hi Joey!
I can only speak for our [2018 paper code](https://github.com/tizian/layer-laboratory/tree/master) but I think in principle it should be feasible. A few comments:
- Most of the adding-doubling machinery (implemented in `liblayer` back then) is independent of Mitsuba 2/3 Enoki/DrJIT code. It relies on a combination of the Eigen/FFTW/UMFPACK libraries that need to be re-added.
- Mitsuba 3 ditched tbb in favor of nanothreads, so some of the parallelized code might need small adjustments.
- The biggest difference will be in handling the actual evaluation of the 2D Fourier series to evaluate the layered BSDFs (e.g. `eval_fourier_series_2d` in `fourier.cpp`). That used a fairly specific way of using Enoki packages to vectorize the code. I think this part of Enoki didn't survive the transition to DrJIT, but maybe something clever can be done with the new JIT compilation to emit fast vectorized code automatically. Worst case, a non-vectorized evaluation would be easy to implement---just a bit slower to run.

If you mostly care about a general layered material, and not necessarily this Fourier-based adding-doubling flavor of it, it might make more sense however to work on a nice reference implementation of the [position-free MC layering](https://shuangz.com/projects/layered-sa18/) in Mitsuba 3. That is quite a bit more flexible in practice (though generally has higher variance). That would (IIUC) also be extendible to Ibón's pearlescent materials at the same time then.
